python/ray/autoscaler/command_runner.py

```diff
@@ -45,7 +47,12 @@ def __init__(self,
 def _with_interactive(cmd):
     force_interactive = ("true && source ~/.bashrc && "
                          "export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignore && ")
-    return ["bash", "--login", "-c", "-i", quote(force_interactive + cmd)]
+    hide_ttys = " 2> >(tee 1>&2) | tee"
```

Comment:
This definitely won't work on Windows :( also it's difficult to understand what the intent is, though I understand you're trying to make TTYs look like regular pipes. What is the actual problem this is solving? There must be a better solution.

Reply:
Nice catch, I didn't even consider whether this works on Windows. But, as you noted below the current `bash -i` wrapping doesn't work anyway, so I guess this is a non-issue lol?

This hack for hiding the TTYs is necessary because we rely on some extremely stupid bash mechanisms to setup our env. Specifically, we want `.bashrc` to execute, which will only happen in interactive mode, AND if you source it manually it will immediately cancel itself. This means we HAVE to start in full-on interactive mode (which we do by launching a login session for some reason???????) including bash and SSH settings (see default SSH flags). BUT this causes apps to print an extreme amount of progress bars, "\r"-based stuff, control codes, etc. This was fine when we just dumped them to a local TTY, but now that we both try to parse the output and write it to file, we end up with `\ESC` gibberish and literally hundreds of `13% |====>-------| (5MB/s)` lines from pip downloading packages.

So, for backwards-compatibility reasons (not changing the env), we keep the interactive session, but we kill the tty-detection in the command by piping everything.

If this seems like an extremely weird way to run ssh commands, then yeah no kidding, hence the workaround is also extremely weird. The interactive session hack also causes the bash "job control is not available" message in the logs (cause login sessions not started by the `login` command are actually all subtly invalid and broken).

P.S. I actually suspect that we depend on a Deep-Learning-AMI-specific file to execute and not `.bashrc`, though I've only been told about `.bashrc`; anyway this does not matter because we still need to preserve backwards compatibility

P.P.S. I think the part of the DLAMI file that we need does some extremely hacky and inappropriate conda setup but we nevertheless depend on it and there seems to be no clean workaround